### PR TITLE
[PATCH v6] example: script: enhance application run script

### PR DESCRIPTION
**Pull request by SunilKumarKori**
Currently example applications uses pcap based pktio to validate the
functionality on linux-generic platform during make check but other
platforms may not be supporting pcap interfaces as pktio.

So example application's run scripts are enhanced to use generic framework
to define pktio interfaces and corresponding interface environment scripts
are placed under platform implementation.

Signed-off-by: Sunil Kumar Kori <skori@marvell.com>

### CODE REVIEW

**Comment by jerinjacobk** [comment]
@psavol @matiaselo @JannePeltonen 
It will helpful if we get a top-level comments on the direction of this patch. This patch really enables another ODP implementation(s) to have a complete pktio validation test case using a generic scheme thus enabling removing unnecessary changes in the example and validation area for other implementations.

**Comment by SunilKumarKori** [comment]
@matiaselo 
I am getting a build error but I am not sure that it because of my patch. It looks something else because the same test case is passed on my local machine. Can you please help me to get it resolved ?

**Comment by matiaselo** [comment]
> @matiaselo
> I am getting a CI error but I am not sure that it because of my patch. It looks something else because the same test case is passed on my local machine. Can you please help me to get it resolved ?
@SunilKumarKori Yeah, traffic manager was once more the culprit. I restarted the test.

**Comment by SunilKumarKori** [comment]
> > I am getting a CI error but I am not sure that it because of my patch. It looks something else because the same test case is passed on my local machine. Can you please help me to get it resolved ?
> > @SunilKumarKori Yeah, traffic manager was once more the culprit. I restarted the test.

@matiaselo Thanks, So should I need to do anything to run CI again ?

**Comment by matiaselo** [comment]
> > > I am getting a CI error but I am not sure that it because of my patch. It looks something else because the same test case is passed on my local machine. Can you please help me to get it resolved ?
> 
> > > @SunilKumarKori Yeah, traffic manager was once more the culprit. I restarted the test.
> 
> 
> 
> Thanks, So should I need to do anything to run CI again ?

Nope, the test runs automatically.

**Comment by matiaselo** [comment]
Hi @SunilKumarKori,

I really like what this PR is doing. Some high level comments. I would prefer that the `interface_env` files are stored under `platform/linux-generec/test`, instead of the new `scripts` directory since these files are for testing purposes. Also, for naming consistency `interface_env` could be renamed to `pktio_env`.

File: example/l2fwd_simple/l2fwd_simple_run.sh [hunk 1]
```diff
@@ -6,32 +6,34 @@
...
-  echo "File sizes disagree"
-  exit 1
-fi
-
-rm -f pcapout.pcap
+validate_result
 
-./odp_l2fwd_simple${EXEEXT} null:0 null:1 \
-	02:00:00:00:00:01 02:00:00:00:00:02 -t 2
+./odp_l2fwd_simple${EXEEXT} $IF2 $IF3 02:00:00:00:00:01 02:00:00:00:00:02 -t 2
```
Comment (matiaselo) [inline example/l2fwd_simple/l2fwd_simple_run.sh:29]: This null interface test doesn't make much sense, so it could be removed.

File: example/l2fwd_simple/l2fwd_simple_run.sh [hunk 2]
```diff
@@ -6,32 +6,34 @@
...
 #
 
-PCAP_IN=`find . ${TEST_DIR} $(dirname $0) -name udp64.pcap -print -quit`
-echo "using PCAP_IN = ${PCAP_IN}"
+if  [ -f ./interface_env ]; then
+	. ./interface_env
+else
+        echo "BUG: unable to find interface_env!"
+        echo "interface_env has to be in current directory"
+        exit 1
```
Comment (matiaselo) [inline example/l2fwd_simple/l2fwd_simple_run.sh:14]: Indentation not matching.

File: example/l3fwd/odp_l3fwd_run.sh [hunk 3]
```diff
@@ -6,23 +6,27 @@
...
+        echo "BUG: unable to find interface_env!"
+        echo "interface_env has to be in current directory"
+        exit 1
+fi
+
+setup_interfaces
 
-./odp_l3fwd${EXEEXT} -i pcap:in=${PCAP_IN},pcap:out=${PCAP_OUT} \
-	    -r "10.0.0.0/24,pcap:out=${PCAP_OUT}" -d 30
+./odp_l3fwd${EXEEXT} -i $IF0,$IF1 -r "10.0.0.0/24,pcap:out=${PCAP_OUT}" -d 30
```
Comment (matiaselo) [inline example/l3fwd/odp_l3fwd_run.sh:19]: This is still using linux-generic specific interface name `pcap:out=${PCAP_OUT}`.

File: example/packet/packet_dump_run.sh [hunk 4]
```diff
@@ -6,15 +6,23 @@
...
-./odp_packet_dump${EXEEXT} -i pcap:in=${PCAP_IN}:loops=10 -n 10 -o 0 -l 64
+./odp_packet_dump${EXEEXT} -i $IF0 -n 10 -o 0 -l 64
 STATUS=$?
 if [ "$STATUS" -ne 0 ]; then
   echo "Error: status was: $STATUS, expected 0"
   exit 1
 fi
 
+cleanup_interfaces
+
```
Comment (matiaselo) [inline example/packet/packet_dump_run.sh:27]: In addition to this one there is `packet/pktio_run.sh` script.

File: example/ping/ping_run.sh [hunk 5]
```diff
@@ -6,21 +6,29 @@
...
+        echo "BUG: unable to find interface_env!"
+        echo "interface_env has to be in current directory"
+        exit 1
+fi
+
+setup_interfaces
 
 # Ping test with 100 ICMP echo request packets (verbose mode)
-./odp_ping${EXEEXT} -v -n 100 -ipcap:in=${PCAP_IN}:out=${PCAP_OUT}
+./odp_ping${EXEEXT} -v -n 100 -ipcap:$IF0:$IF1
```
Comment (matiaselo) [inline example/ping/ping_run.sh:20]: Implementation specific `pcap` prefix.

File: example/switch/switch_run.sh [hunk 6]
```diff
@@ -6,33 +6,27 @@
...
 
-RX_PORTS=""
-for i in `seq 1 $NUM_RX_PORT`;
-do
-	RX_PORTS="${RX_PORTS},pcap:out=pcapout${i}.pcap"
-done
+setup_interfaces
 
-./odp_switch${EXEEXT} -i pcap:in=${PCAP_IN}${RX_PORTS} -t 1
+./odp_switch${EXEEXT} -i $IF0$IF1 -t 1
```
Comment (matiaselo) [inline example/switch/switch_run.sh:21]: Maybe it's clearer just to use `$IF0,$IF1,$IF2,$IF3` here.

**Comment by matiaselo** [comment]
`test/performance/odp_pktio_ordered_run.sh` is also using pcap pktio.

File: platform/linux-generic/Makefile.am [hunk 7]
```diff
@@ -4,6 +4,8 @@
 include $(top_srcdir)/platform/Makefile.inc
 lib_LTLIBRARIES += $(LIB)/libodp-linux.la
 
+SUBDIRS = scripts
```
Comment (matiaselo) [inline platform/linux-generic/Makefile.am:7]: I'd prefer storing the files under `platform/linux-generic/test`.

File: platform/linux-generic/scripts/example/l2fwd_simple/interface_env [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#!/bin/sh
+#
+# Copyright (c) 2015-2018, Linaro Limited
```
Comment (matiaselo) [inline platform/linux-generic/scripts/example/l2fwd_simple/interface_env:3]: Marvell copyright missing.

File: example/l2fwd_simple/l2fwd_simple_run.sh (hunk 1 above)
Comment (SunilKumarKori) [reply example/l2fwd_simple/l2fwd_simple_run.sh:29]: Ack.

File: example/l2fwd_simple/l2fwd_simple_run.sh (hunk 2 above)
Comment (SunilKumarKori) [reply example/l2fwd_simple/l2fwd_simple_run.sh:14]: Okay, I will check.

File: example/l3fwd/odp_l3fwd_run.sh (hunk 3 above)
Comment (SunilKumarKori) [reply example/l3fwd/odp_l3fwd_run.sh:19]: Missed to replace. Will take in next version.

File: example/packet/packet_dump_run.sh (hunk 4 above)
Comment (SunilKumarKori) [reply example/packet/packet_dump_run.sh:27]: Ack.

File: example/ping/ping_run.sh (hunk 5 above)
Comment (SunilKumarKori) [reply example/ping/ping_run.sh:20]: Ack.

File: example/switch/switch_run.sh (hunk 6 above)
Comment (SunilKumarKori) [reply example/switch/switch_run.sh:21]: Yeah. I thought so but just to use existing one, I followed this one. I will change in next version.

File: platform/linux-generic/Makefile.am (hunk 7 above)
Comment (SunilKumarKori) [reply platform/linux-generic/Makefile.am:7]: Ack.

File: platform/linux-generic/scripts/example/l2fwd_simple/interface_env (hunk 8 above)
Comment (SunilKumarKori) [reply platform/linux-generic/scripts/example/l2fwd_simple/interface_env:3]: Ack.

**Comment by SunilKumarKori** [comment]
@matiaselo It looks like CI has faced the same issue again (traffic manager issue). Can you please check and rerun the test ?

**Comment by matiaselo** [comment]
> It looks like CI has faced the same issue again (traffic manager issue). Can you please check and rerun the test ?
@SunilKumarKori Done. I'll try to go through the updated PR tomorrow morning.

**Comment by SunilKumarKori** [comment]
> > It looks like CI has faced the same issue again (traffic manager issue). Can you please check and rerun the test ?
> > @SunilKumarKori Done. I'll try to go through the updated PR tomorrow morning.

Sure, Thanks.

**Comment by matiaselo** [comment]
Hi @SunilKumarKori,

Sorry for taking so long with the review.

Instead of using e.g. platform/linux-generic/test/**scripts**/ping/ to store the new files, I'd prefer using platform/linux-generic/test/**example**/ping/. This way the folder structure matches the top level.

Also, after making this change the current `platform/linux-generic/test/ipsec` directory could be moved under `platform/linux-generic/test/example`. This should be done in a separate commit.

**Comment by matiaselo** [comment]
Not directly related to this PR, but @JannePeltonen noticed that since Stan's 75b0ae386 commit, `make distcheck` doesn't work anymore without sudo.

E.g.
```
make[3]: Entering directory '/home/nnn/dev/odp.git/opendataplane-1.24.2.0/_build/sub/platform/linux-generic/test/pktio_ipc'
if [ "x../../../../../../platform/linux-generic/test/pktio_ipc" != "x." ]; then \
	for f in pktio_ipc_run.sh; do \
		if [ -e ../../../../../../platform/linux-generic/test/pktio_ipc/$f ]; then \
			mkdir -p ./$(dirname $f); \
			cp ../../../../../../platform/linux-generic/test/pktio_ipc/$f ./$f; \
		fi \
	done \
fi
cp: cannot create regular file './pktio_ipc_run.sh': Permission denied
Makefile:995: recipe for target 'all-local' failed
make[3]: *** [all-local] Error 1
make[3]: Leaving directory '/home/nnn/dev/odp.git/opendataplane-1.24.2.0/_build/sub/platform/linux-generic/test/pktio_ipc'
Makefile:803: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/home/nnn/dev/odp.git/opendataplane-1.24.2.0/_build/sub/platform/linux-generic/test'
Makefile:565: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/nnn/dev/odp.git/opendataplane-1.24.2.0/_build/sub'
Makefile:776: recipe for target 'distcheck' failed
make: *** [distcheck] Error 1
```

It would be much appreciated if you could fix this while you are touching to the Makefiles anyway.

File: platform/linux-generic/test/scripts/switch/pktio_env [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		if [ `stat -c %s pcapout${i}.pcap` -ne `stat -c %s ${PCAP_IN}` ]; then
+			echo "Error: Output file $i size not matching"
+		fi
+		rm -f pcapout${i}.pcap
+	done
+}
+
+setup_interfaces()
+{
+	echo "pktio: setting up test interfaces $IF0 $IF1."
```
Comment (matiaselo) [inline platform/linux-generic/test/scripts/switch/pktio_env:45]: Uses 4 interfaces.

File: platform/linux-generic/test/scripts/switch/pktio_env [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	return 0
+}
+
+cleanup_interfaces()
+{
+	echo "pktio: cleaning up test interfaces $IF0 $IF1."
```
Comment (matiaselo) [inline platform/linux-generic/test/scripts/switch/pktio_env:51]: Same as above.

File: platform/linux-generic/test/scripts/l2fwd_simple/pktio_env [hunk 11]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+#
+# Copyright(C) 2020 Marvell.
```
Comment (matiaselo) [inline platform/linux-generic/test/scripts/l2fwd_simple/pktio_env:3]: A minor issue but it's better to keep the copyright message format consistent across the whole project.
So `# Copyright (c) 2020, Marvell`

Comment (SunilKumarKori) [reply platform/linux-generic/test/scripts/l2fwd_simple/pktio_env:3]: Ack.

File: platform/linux-generic/test/scripts/switch/pktio_env (hunk 10 above)
Comment (SunilKumarKori) [reply platform/linux-generic/test/scripts/switch/pktio_env:51]: Actually I included all three interfaces into single but I think it will be better to keep them separate.

**Comment by SunilKumarKori** [comment]
@matiaselo Following are the disposition:

1. Regarding test/performance/odp_pktio_ordered_run.sh, my opinion was to send it as a separate PR because first I was targeting to examples. Once it will be done I will send for next PR to handle **odp_pktio_ordered_run.sh** Please suggest. 
2. I will Makefile changes to handle make distclean issue.

**Comment by matiaselo** [comment]
> 1. Regarding test/performance/odp_pktio_ordered_run.sh, my opinion was to send it as a separate PR because first I was targeting to examples. Once it will be done I will send for next PR to handle **odp_pktio_ordered_run.sh** Please suggest.

OK

> 2. I will do Makefile changes to handle _make distclean_ issue.

Thanks!

File: platform/linux-generic/test/scripts/switch/pktio_env (hunk 9 above)
Comment (SunilKumarKori) [reply platform/linux-generic/test/scripts/switch/pktio_env:45]: Ack.

**Comment by SunilKumarKori** [comment]
@matiaselo CI failed again.  Can you please look into this ?

**Comment by matiaselo** [comment]
> CI failed again. Can you please look into this ?
I restarted the test.

**Comment by SunilKumarKori** [comment]
> > CI failed again. Can you please look into this ?
> 
> I restarted the test.

Thanks.

**Comment by SunilKumarKori** [comment]
@matiaselo CI tests are passed. Please have a look on PR once you get the time. Thanks.

File: example/generator/Makefile.am [hunk 12]
```diff
@@ -21,7 +21,7 @@ all-local:
 		for f in $(EXTRA_DIST); do \
 			if [ -e $(srcdir)/$$f ]; then \
 				mkdir -p $(builddir)/$$(dirname $$f); \
-				cp $(srcdir)/$$f $(builddir)/$$f; \
+				cp -f $(srcdir)/$$f $(builddir)/$$f; \
```
Comment (matiaselo) [inline example/generator/Makefile.am:24]: Please split the distcheck fix into a separate commit.

File: example/simple_pipeline/simple_pipeline_run.sh [hunk 13]
```diff
@@ -9,27 +9,32 @@
 # Exit code expected by automake for skipped tests
 TEST_SKIPPED=77
 
-PCAP_IN=`find . ${TEST_DIR} $(dirname $0) -name udp64.pcap -print -quit`
-echo "using PCAP_IN = ${PCAP_IN}"
+# Use examples_env for make check from platform directory or locally installed
```
Comment (matiaselo) [inline example/simple_pipeline/simple_pipeline_run.sh:12]: This comment should probably be removed?

File: platform/linux-generic/test/example/l2fwd_simple/pktio_env [hunk 14]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+#
+# Copyright (C) 2020 Marvell.
```
Comment (matiaselo) [inline platform/linux-generic/test/example/l2fwd_simple/pktio_env:3]: For consistency please use this style for all copyright lines:
`Copyright (c) 2020, Marvell`

**Comment by matiaselo** [comment]
@SunilKumarKori I had three final comments (split commit, copyright messages, simple_pipeline_run.sh) but otherwise the PR is looking good to me. You can add my reviewed-by tag to the commits.

We just merged some patches so you should also do a rebase. However, there shouldn't be any conflicts.

**Comment by SunilKumarKori** [comment]
@matiaselo Updated PR with suggested comments. Please do the needful once CI gets passed.

**Review by matiaselo — APPROVED** [review]

**Comment by matiaselo** [comment]
@SunilKumarKori Looking good, just add my tag `Reviewed-by: Matias Elo <matias.elo@nokia.com>` to the commit messages and I can merge this PR after the tests pass.

**Comment by SunilKumarKori** [comment]
> @SunilKumarKori Looking good, just add my tag `Reviewed-by: Matias Elo <matias.elo@nokia.com>` to the commit messages and I can merge this PR after the tests pass.

Done.

**Comment by SunilKumarKori** [comment]
@matiaselo It looks ci is stuck .

**Comment by matiaselo** [comment]
> It looks ci is stuck .

There are other PRs in the CI queue before this one. The tests should done tomorrow morning.
